server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java

```diff
@@ -0,0 +1,115 @@
+    }
+
+    public void validate(DesiredNodes desiredNodes) {
+        final List<Tuple<Integer, RuntimeException>> exceptions = new ArrayList<>();
+        final List<DesiredNode> nodes = desiredNodes.nodes();
+        for (int i = 0; i < nodes.size(); i++) {
+            final DesiredNode node = nodes.get(i);
+            try {
+                validate(node);
+            } catch (RuntimeException e) {
+                exceptions.add(Tuple.tuple(i, e));
+            }
+        }
+
+        if (exceptions.isEmpty() == false) {
+            final String nodeIndicesWithFailures = exceptions.stream()
+                .map(Tuple::v1)
+                .map(i -> Integer.toString(i))
+                .collect(Collectors.joining(","));
+            IllegalArgumentException invalidSettingsException = new IllegalArgumentException(
+                format(Locale.ROOT, "Nodes in positions [%s] contain invalid settings", nodeIndicesWithFailures)
+            );
+            for (Tuple<Integer, RuntimeException> exceptionTuple : exceptions) {
+                invalidSettingsException.addSuppressed(exceptionTuple.v2());
+            }
+            throw invalidSettingsException;
+        }
+    }
+
+    private void validate(DesiredNode node) {
+        if (node.externalId() == null) {
+            throw new IllegalArgumentException(format(Locale.ROOT, "[%s] is missing or empty", NODE_EXTERNAL_ID_SETTING.getKey()));
+        }
+
+        final Settings settings = node.settings();
+
+        final Set<Setting<?>> settingSet = new HashSet<>();
+        final List<String> unknownSettings = new ArrayList<>();
+        // If we're dealing with a node in a newer version it's possible
+        // that this node doesn't know about these. In that case we do a best
+        // effort and try to validate the settings that are known to this version.
+        final Settings.Builder updatedSettingsBuilder = Settings.builder().put(settings);
+        boolean hasOverriddenSettings = false;
+        for (String settingName : settings.keySet()) {
+            Setting<?> setting = clusterSettings.get(settingName);
+            if (setting == null) {
+                unknownSettings.add(settingName);
+                updatedSettingsBuilder.remove(settingName);
+            } else {
+                // Some settings might be defined using environmental information
+                // such as the number of available processors to define valid setting
+                // value ranges, in that case we should define new settings that take
+                // into account the node properties that we're evaluating.
+                Setting<?> maybeOverrideSetting = maybeOverride(node, setting);
+                hasOverriddenSettings |= (setting != maybeOverrideSetting);
+                settingSet.add(maybeOverrideSetting);
+            }
+        }
+
+        if (unknownSettings.isEmpty() == false && node.version().onOrBefore(Version.CURRENT)) {
```

Comment:
I think we would want something like the version barrier here, i.e., not allow going backwards in node versions. Can certainly wait for a follow-up.